apps/passport-client/src/localstorage.ts

```diff
@@ -12,11 +12,22 @@ import { SemaphoreSignaturePCD } from "@pcd/semaphore-signature-pcd";
 import { Identity } from "@semaphore-protocol/identity";
 import { z } from "zod";
 import { getPackages } from "./pcdPackages";
+import {
+  logAndUploadValidationErrors,
+  validatePCDCollection
+} from "./validateState";
+
+// validation points in this file
 
 const OLD_PCDS_KEY = "pcds"; // deprecated
 const COLLECTION_KEY = "pcd_collection";
 
 export async function savePCDs(pcds: PCDCollection): Promise<void> {
+  const validationErrors = validatePCDCollection(pcds);
+  if (validationErrors.length > 0) {
+    logAndUploadValidationErrors(validationErrors);
+    throw new Error("couldn't save PCDs\n:" + validationErrors.join("\n"));
```

Comment:
See my comment below about error reporting.  Hopefully this line keeps bad states from being saved, so maybe the user doesn't need to log off to recover in this case?